### ISSUE TITLE
WebView doesn't close on extension (but not window) reload

### ISSUE THREAD

**Issue by SchoofsKelvin** [Microsoft/vscode#70656]
### Summary
Some actions can cause extensions to be reloaded without the whole window being reloaded. This results in WebViews opened by extensions to remain open. On its own not a problem, except should these use the message API to "talk" to their extension, since it got reloaded/terminated. Expected behavior is to close all WebViews, similar to what happens when the window is reloaded, e.g. using the `Developer: Reload Window` command.

### Steps to reproduce on your own
1. Have an extension launch a WebView
2. Change the "main workspace folder" without triggering a window reload
I discovered this by my extension using `updateWorkspaceFolders` to go from 0 to 1 folders.
Having a multi-folder workspace, rightclicking the first folder, and using `Remove Folder from Workspace` also triggers this effect. So does pressing `Reload` after installing a (new) .vsix for an already installed extension, but that results in an empty WebView?
3. The WebView remains, working, but the extensions reloaded, "breaking" `onDidReceiveMessage`

### With [example extension](https://drive.google.com/file/d/19O86YV-_IWptsLu-894k6fEfvWf-fHmK/view) to easily replicate
*(source code available by opening the file as an archive, or doing step 4 below)*
1. Open a new window (and install the extension if you haven't yet)
2. Run the `ISSUE: Open a test WebView` command
3. Click `Send message` and notice how you get a response from the extension
4. Run the `ISSUE: Add/remove the first workspace folder to trigger the semi-reload` command
If you had folders in your workspace, the first one gets removed. If your workspace is empty, it'll add the extension's folder (with source code) as workspace folder.
5. Click `Send message` and notice how you don't get a response anymore

_____

- VSCode versions tested: 
	- 1.32.3 (user setup) (a3db5be9)
	- 1.33.0-insider (user setup) (5a88160c)
- OS Version: Windows 10 Version 1803 (OS Build 17134.407)

> If the first workspace folder is added, removed or changed, the currently executing extensions (including the one that called this method) will be terminated and restarted so that the (deprecated) rootPath property is updated to point to the first workspace folder.
***From the documentation of vscode.workspace.updateWorkspaceFolders***

**Comment by mjbvz** [comment]
We do not automatically destroy webviews on extension reloads as not all webviews require their backing extension to stick around. Does your extension have a `deactivate` method that calls dispose on the webviews it creates or add its webview to the `context.subscriptions` list?

**Comment by SchoofsKelvin** [comment]
I haven't. I've seen `dispose()` being mentioned in the [Webview API guide](https://code.visualstudio.com/api/extension-guides/webview), but it didn't mention anything about disposing-on-extension-reload in regards to `onDidReceiveMessage`.

As an alternative, I've been trying to use the WebviewPanelSerializer, but with no luck:
```ts
  // inside my exported activate method
  vscode.window.registerWebviewPanelSerializer('sshfs-settings', {
    async deserializeWebviewPanel(panel, state) {
      // Breakpoint here, although I've also tried console.log to be safe
      const { initiateWebviewPanel } = await import('./settings');
      await initiateWebviewPanel(panel, state);
    },
  });
```
The deserializing only happens on window reload, but **not** when extensions get reloaded (by changing the first WorkspaceFolder), although reloading the window afterwards does trigger the deserializer. Although me not disposing of the webview during `deactivate` isn't a bug then, the WebView not being deserialized on extension reload might be?

Perhaps always triggering the deserializer, or having it check whether a deserializer is registered *(or a deserializeOnReload flag in WebviewOptions)* might be helpful here. Right now, when the first WorkspaceFolder changes, the WebviewPanel would have to be disposed (which would be weird towards the user), since after the extension reloads, there's no way to "reconnect" the Webview from the extension.

**Comment by mjbvz** [comment]
Yes, you can use `dispose` if you want to clean up webviews on extension unload and then re-create them on reload.

We should probably support serialization on extension reload, although it needs some thought

**Comment by bencefr** [comment]
We face the same issue in our extension and I can add to this that even if I try to explicitly dispose the webviews on deactivate  it does not work, since deactivate itself is not called in this case. A new instance of the extension is activated without the old being deactivated!

**Comment by thomasAtToem** [comment]
We have a similar problem with a custom editor. 
The extension get reloaded after adding a folder resulting into new workspace.
After deactivation /activation of the extension has NO access to the editor but the webview is still open.
Closing the editor on deactivate is no solution as it shall stay open like on window reload.

To reproduce 

- Open the PawDraw editor from the extensions.
- Add folder to workspace (before there is only a folder open - no workspace).
- Paw draw visible but not functional
 
thomas

**Comment by benjamin-t-santos** [comment]
Is there any update on this issue?
My webivew panel is still not disposed properly when a new folder is added to the workspace. Even when I try and manually dispose of it during `deactivate()`.
Using console.log + the debugger, I see upon adding a new folder to the workspace `deactivate()` is called, in which I explicitly call `dispose()` on my webview panel. But the webview remains open and the webview panel is left dangling.

**Comment by juancampa** [comment]
We're also experiencing this problem. Related issue https://github.com/microsoft/vscode/issues/188257

Indeed disposing a webview during `deactivate` does not actually dispose it unfortunately. `console.log` also sometimes doesn't so it looks like `deactivate` does't get called but it does.

**Comment by juancampa** [comment]
I was hoping I could work around it by sending pings from the extension to the webview, and having the webview close itself when no pings are received but `window.close()` doesn't work either :(
